File: cpp/source/vec3.cpp
struct vec3 {
	float x, y, z;
};

inline vec3 operator+( const vec3 &a, const vec3 &b ) {
	vec3 result = {
		a.x + b.x,
		a.y + b.y,
		a.z + b.z
	};

	return result;
}

inline vec3 operator-( const vec3 &a ) {
	vec3 result = {
		-a.x,
		-a.y,
		-a.z
	};

	return result;
}

inline vec3 operator-( const vec3 &a, const vec3 &b ) {
	vec3 result = {
		a.x - b.x,
		a.y - b.y,
		a.z - b.z
	};

	return result;
}

inline vec3 operator*( const vec3 &a, const vec3&b ) {
	vec3 result = {
		a.x * b.x,
		a.y * b.y,
		a.z * b.z
	};

	return result;
}

inline vec3 operator*( const vec3 &a, float b ) {
	vec3 result = {
		a.x * b,
		a.y * b,
		a.z * b
	};

	return result;
}

inline vec3 operator*( float b, const vec3 &a ) {
	vec3 result = {
		a.x * b,
		a.y * b,
		a.z * b
	};

	return result;
}

inline vec3 operator/( const vec3 &a, double b ) {
	return (1/b) * a;
}

inline float dot( const vec3 &a, const vec3 &b ) {
	return a.x * b.x + a.y * b.y + a.z * b.z;
}

inline vec3 cross( const vec3 u, const vec3 &v ) {
	vec3 result = {
		u.y * v.z - u.z * v.y,
		u.z * v.x - u.x * v.z,
		u.x * v.y - u.y * v.x
	};

	return result;
}

inline float length2( const vec3 &a ) {
	return dot( a, a );
}

inline float length( const vec3 &a ) {
	return sqrt( length2( a ) );
}

inline vec3 normalize( vec3 a ) {
	return a * ( 1.0 / length( a ) );
}

vec3 vec3_random() {
	vec3 result = {
		random(),
		random(),
		random()
	};

	return result;
}

vec3 vec3_random( float min, float max ) {
	vec3 v_min = { min, min, min };
	return ( max - min ) * vec3_random() + v_min;
}

File: web/script/main/vec3.js
"use strict";

function vec3( x, y, z ) {
	this.x = x;
	this.y = y;
	this.z = z;
}

vec3.add = function( a, b, c, d ) {
	if ( d ) return new vec3(
		a.x + b.x + c.x + d.x,
		a.y + b.y + c.y + d.y,
		a.z + b.z + c.z + d.z
	);

	if ( c ) return new vec3(
		a.x + b.x + c.x,
		a.y + b.y + c.y,
		a.z + b.z + c.z
	);

	return new vec3(
		a.x + b.x,
		a.y + b.y,
		a.z + b.z
	);
};

vec3.neg = function( a ) {
	return new vec3(
		-a.x,
		-a.y,
		-a.z
	);
};

vec3.sub = function( a, b, c, d ) {
	if ( d ) return new vec3(
		a.x - b.x - c.x - d.x,
		a.y - b.y - c.y - d.y,
		a.z - b.z - c.z - d.z
	);

	if ( c )return new vec3(
		a.x - b.x - c.x,
		a.y - b.y - c.y,
		a.z - b.z - c.z
	);

	return new vec3(
		a.x - b.x,
		a.y - b.y,
		a.z - b.z
	);
};

vec3.mul = function( a, t ) {
	return new vec3( a.x * t, a.y * t, a.z * t );
};

// vec3.prototype.mul = function( t ) {
// 	return new vec3( this.x * t, this.y * t, this.z * t );
// };

vec3.div = function( a, b ) {
	return vec3.mul( a, 1 / b );
};

vec3.dot = function( a, b ) {
	return a.x * b.x + a.y * b.y + a.z * b.z;
};

vec3.cross = function( u, v ) {
	return new vec3(
		u.y * v.z - u.z * v.y,
		u.z * v.x - u.x * v.z,
		u.x * v.y - u.y * v.x
	);
};

vec3.length2 = function( a ) {
	return vec3.dot( a, a );
};

vec3.normalize = function( a ) {
	return vec3.mul( a, 1 / Math.sqrt( vec3.length2( a ) ) );
};

vec3.random = function( min = null, max = null ) {
	if ( min == null ) return new vec3( random(), random(), random() );
	let v_min = new vec3( min, min, min );
	return ( max - min ) * vec3_random() + v_min;
};

vec3.mix = function( a, b, t ) {
	return vec3.add( vec3.mul( a, t ), vec3.mul( b, 1 - t ) );
};

export default vec3;
